Windows build broken

The [last commit](https://github.com/Byron/open-rs/commit/88ddb6febe811fa8f5ab12b32cbf2a716676fb53) broke the build on Windows:

```
error[E0308]: match arms have incompatible types                         ] 0/2: open
  --> src\lib.rs:68:13
   |
68 | /             match path_ref.to_str() {
69 | |                 Some(s) => s.replace("&", "^&"),
70 | |                 None => path_ref,
   | |                         -------- match arm with an incompatible type
71 | |             }
   | |_____________^ expected struct `std::string::String`, found reference
   |
   = note: expected type `std::string::String`
              found type `&std::ffi::OsStr`
```

I'm planning on sending a PR that replaces the Windows code with [the approach rustup uses](https://github.com/rust-lang-nursery/rustup.rs/blob/22b6cdb8b000f9217af1dd5dd4e73ce032bdb25d/src/rustup-utils/src/raw.rs#L390-L424) anyways, which will take care of this issue too.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -13,3 +13,6 @@ documentation = "http://byron.github.io/open-rs"
 test = false
 doc = false
 name = "open"
+
+[target.'cfg(windows)'.dependencies]
+winapi = { version = "0.3", features = ["shellapi"] }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -37,9 +37,15 @@
 //! }
 //! # }
 //! ```
+#[cfg(windows)]
+extern crate winapi;
+
+#[cfg(not(target_os = "windows"))]
+use std::process::Command;
+
 use std::ffi::OsStr;
 use std::io;
-use std::process::{Command, ExitStatus};
+use std::process::ExitStatus;
 
 #[cfg(not(any(target_os = "windows", target_os = "macos")))]
 pub fn that<T: AsRef<OsStr> + Sized>(path: T) -> io::Result<ExitStatus> {
@@ -59,22 +65,53 @@ pub fn that<T: AsRef<OsStr> + Sized>(path: T) -> io::Result<ExitStatus> {
 
 #[cfg(target_os = "windows")]
 pub fn that<T: AsRef<OsStr> + Sized>(path: T) -> io::Result<ExitStatus> {
-    Command::new("cmd")
-        .arg("/C")
-        .arg("start")
-        .arg("")
-        .arg({
-            let path_ref = path.as_ref();
-            match path_ref.to_str() {
-                Some(s) => s.replace("&", "^&"),
-                None => path_ref,
-            }
-        })
-        .spawn()?
-        .wait()
+    use winapi::ctypes::c_int;
+    use winapi::um::shellapi::ShellExecuteW;
+    use std::os::windows::ffi::OsStrExt;
+    use std::os::windows::process::ExitStatusExt;
+    use std::ptr;
+
+    const SW_SHOW: c_int = 5;
+
+    let path = windows::convert_path(path.as_ref())?;
+    let operation: Vec<u16> = OsStr::new("open\0").encode_wide().collect();
+    let result = unsafe {
+        ShellExecuteW(
+            ptr::null_mut(),
+            operation.as_ptr(),
+            path.as_ptr(),
+            ptr::null(),
+            ptr::null(),
+            SW_SHOW,
+        )
+    };
+    if result as c_int > 32 {
+        Ok(ExitStatus::from_raw(0))
+    } else {
+        Err(io::Error::last_os_error())
+    }
 }
 
 #[cfg(target_os = "macos")]
 pub fn that<T: AsRef<OsStr> + Sized>(path: T) -> io::Result<ExitStatus> {
     Command::new("open").arg(path.as_ref()).spawn()?.wait()
+}
+
+#[cfg(windows)]
+mod windows {
+    use std::io;
+    use std::ffi::OsStr;
+    use std::os::windows::ffi::OsStrExt;
+
+    pub fn convert_path(path: &OsStr) -> io::Result<Vec<u16>> {
+        let mut maybe_result: Vec<_> = path.encode_wide().collect();
+        if maybe_result.iter().any(|&u| u == 0) {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "path contains NUL byte(s)",
+            ));
+        }
+        maybe_result.push(0);
+        Ok(maybe_result)
+    }
 }

